src/languages/haskell.js

```diff
@@ -7,8 +7,26 @@ Category: functional
 */
 
 export default function(hljs) {
+
+  /* See:
+     - https://www.haskell.org/onlinereport/lexemes.html
+     - https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/exts/binary_literals.html
+     - https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/exts/numeric_underscores.html
+     - https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/exts/hex_float_literals.html
+  */
+  const decimalDigits = '([0-9]_*)+';
+  const hexDigits = '([0-9a-fA-F]_*)+';
+  const binaryDigits = '([01]_*)+';
+  const octalDigits = '([0-7]_*)+';
+  const ascSymbol = '[!#$%&*+.\\/<=>?@\\\\^~-]';
+  const uniSymbol = '(\\p{S}|\\p{P})' // Symbol or Punctuation
+  const special = '[(),;\\[\\]`|{}]';
+  const symbol = `(${ascSymbol}|(?!(${special}|[_:"']))${uniSymbol})`;
+
   const COMMENT = { variants: [
-    hljs.COMMENT('--', '$'),
+    // Double dash forms a valid comment only if it's not part of legal lexeme.
+    // See: Haskell 98 report: https://www.haskell.org/onlinereport/lexemes.html
+    hljs.COMMENT(`(?<!${symbol})--+(?!${symbol})`, '$'),
```

Comment:
Can't have negative look-behind, not well supported by Safari yet.  Sorry.

Reply:
If we don't need the first character (to be highlighted) the simplest way to do this is include it in the match, but not higlight it... multi-match is good for such things:

Where `yyyy` is a keyword only when NOT preceeded by `x`.

```
{
begin: [
  /[^x]/,
  /yyyy/
],
beginScope: {
  2: "keyword"
},
}
```